File: mower/pico_micropython_ws/firmware/app/protocol.py
# MicroPython-friendly serial protocol utilities: COBS framing + CRC16-CCITT
# Frame format (before COBS):
#   [msg_id(1)][len(1)][seq(1)][ts_ms(4)][payload(len)][crc16(2)]
# After COBS-encoding, a single 0x00 delimiter terminates each frame on the wire.

from micropython import const
# ---- Constants ----
HEADER_SIZE = const(7)   # msg_id(1) + len(1) + seq(1) + ts_ms(4)
CRC_SIZE = const(2)
FRAME_OVERHEAD = const(HEADER_SIZE + CRC_SIZE)

# Message IDs (Pico  Host and Host  Pico)
MSG_IMU = const(0x01)
MSG_ODOM = const(0x02)
MSG_SONAR = const(0x03)
MSG_BATT = const(0x04)
MSG_EVENT = const(0x05)
MSG_BLADES_RPM = const(0x06)

MSG_CMD_DRIVE  = const(0x10)
MSG_CMD_BLADES = const(0x11)
MSG_CMD_RELAY  = const(0x12)
MSG_CMD_LIMITS = const(0x13)
# ---- CRC16-CCITT (poly 0x1021, init 0xFFFF, no xorout) ----
def crc16_ccitt(data: bytes, init: int = 0xFFFF) -> int:
    crc = init
    for b in data:
        crc ^= (b << 8) & 0xFFFF
        for _ in range(8):
            if crc & 0x8000:
                crc = ((crc << 1) ^ 0x1021) & 0xFFFF
            else:
                crc = (crc << 1) & 0xFFFF
    return crc & 0xFFFF

# ---- Minimal COBS encode/decode ----
# Note: We terminate frames with a single zero byte (0x00) after COBS-encoding.
# See: https://en.wikipedia.org/wiki/Consistent_Overhead_Byte_Stuffing

def cobs_encode(data: bytes) -> bytes:
    if not data:
        return b"\x01\x00"
    out = bytearray()
    idx = 0
    while idx < len(data):
        code_idx = len(out)
        out.append(0)  # placeholder for code
        code = 1
        while idx < len(data) and data[idx] != 0 and code < 0xFF:
            out.append(data[idx])
            idx += 1
            code += 1
        out[code_idx] = code
        if idx < len(data) and data[idx] == 0:
            idx += 1  # skip the zero byte
    out.append(0)  # delimiter
    return bytes(out)


def cobs_decode(stream: bytes) -> bytes:
    # stream should be a single COBS-encoded frame WITHOUT the trailing delimiter
    out = bytearray()
    idx = 0
    while idx < len(stream):
        code = stream[idx]
        if code == 0:
            # invalid in COBS body
            return b""
        idx += 1
        end = idx + code - 1
        if end > len(stream):
            # truncated
            return b""
        out.extend(stream[idx:end])
        idx = end
        if code < 0xFF and idx < len(stream):
            out.append(0)
    return bytes(out)

# ---- Packing/unpacking frames ----
import ustruct as _struct
import time


def pack_frame(msg_id: int, payload: bytes, seq: int) -> bytes:
    if payload is None:
        payload = b""
    if len(payload) > 255:
        raise ValueError("payload too large")
    ts_ms = time.ticks_ms() & 0xFFFFFFFF
    header = _struct.pack('<BBBI', msg_id & 0xFF, len(payload) & 0xFF, seq & 0xFF, ts_ms)
    body = header + payload
    crc = crc16_ccitt(body)
    frame = body + _struct.pack('<H', crc)
    return cobs_encode(frame)


class COBSStreamDecoder:
    """Incremental decoder. Feed raw bytes from UART; get full decoded frames (raw body bytes)."""
    def __init__(self, max_frame=512):
        self._buf = bytearray()
        self._frames = []
        self._max = max_frame

    def feed(self, data: bytes):
        if not data:
            return
        for b in data:
            if b == 0:
                # end of frame; decode
                if self._buf:
                    decoded = cobs_decode(self._buf)
                    if decoded:
                        self._frames.append(decoded)
                self._buf = bytearray()
            else:
                if len(self._buf) < self._max:
                    self._buf.append(b)
                else:
                    # overflow: reset buffer
                    self._buf = bytearray()

    def frames(self):
        if not self._frames:
            return []
        out = self._frames
        self._frames = []
        return out


def parse_frame(decoded: bytes):
    """Returns (msg_id, seq, ts_ms, payload) if CRC valid else None."""
    if len(decoded) < FRAME_OVERHEAD:
        return None
    # verify CRC
    body = decoded[:-2]
    rx_crc = (decoded[-2] | (decoded[-1] << 8)) & 0xFFFF
    calc_crc = crc16_ccitt(body)
    if rx_crc != calc_crc:
        return None
    # unpack header
    msg_id, length, seq, ts_ms = _struct.unpack_from('<BBBI', body, 0)
    if HEADER_SIZE + length != len(body):
        return None
    payload = body[HEADER_SIZE:]
    return (msg_id, seq, ts_ms, payload)

File: pico_cpp_ws/firmware/src/protocol.cpp
#include "protocol.hpp"

namespace proto {

uint16_t crc16_ccitt(const uint8_t* data, size_t len, uint16_t init) {
    uint16_t crc = init;
    for (size_t i = 0; i < len; ++i) {
        crc ^= static_cast<uint16_t>(data[i]) << 8;
        for (int b = 0; b < 8; ++b) {
            if (crc & 0x8000) crc = (crc << 1) ^ 0x1021; else crc <<= 1;
        }
    }
    return crc;
}

void cobs_encode(const std::vector<uint8_t>& in, std::vector<uint8_t>& out) {
    out.clear();
    out.reserve(in.size() + 2);
    size_t code_index = 0;
    uint8_t code = 1;
    out.push_back(0); // placeholder
    for (uint8_t byte : in) {
        if (byte == 0) {
            out[code_index] = code;
            code_index = out.size();
            out.push_back(0);
            code = 1;
        } else {
            out.push_back(byte);
            code++;
            if (code == 0xFF) {
                out[code_index] = code;
                code_index = out.size();
                out.push_back(0);
                code = 1;
            }
        }
    }
    out[code_index] = code;
    out.push_back(0x00);
}

void pack_and_encode(uint8_t msg_id, const std::vector<uint8_t>& payload,
                     uint8_t seq, uint32_t ts_ms,
                     std::vector<uint8_t>& encoded) {
    std::vector<uint8_t> buf;
    buf.reserve(7 + payload.size() + 2);
    buf.push_back(msg_id);
    buf.push_back(static_cast<uint8_t>(payload.size()));
    buf.push_back(seq);
    buf.push_back(static_cast<uint8_t>(ts_ms & 0xFF));
    buf.push_back(static_cast<uint8_t>((ts_ms >> 8) & 0xFF));
    buf.push_back(static_cast<uint8_t>((ts_ms >> 16) & 0xFF));
    buf.push_back(static_cast<uint8_t>((ts_ms >> 24) & 0xFF));
    buf.insert(buf.end(), payload.begin(), payload.end());
    uint16_t crc = crc16_ccitt(buf.data(), buf.size());
    buf.push_back(static_cast<uint8_t>(crc & 0xFF));
    buf.push_back(static_cast<uint8_t>((crc >> 8) & 0xFF));
    cobs_encode(buf, encoded);
}

// ---- Decoder side ----
static constexpr size_t HEADER_SIZE = 7; // id(1) len(1) seq(1) ts(4)

bool cobs_decode(const std::vector<uint8_t>& in, std::vector<uint8_t>& out) {
    out.clear();
    size_t idx = 0;
    while (idx < in.size()) {
        uint8_t code = in[idx++];
        if (code == 0) return false; // invalid
        size_t end = idx + code - 1;
        if (end > in.size()) return false; // truncated
        out.insert(out.end(), in.begin() + idx, in.begin() + end);
        idx = end;
        if (code < 0xFF && idx < in.size()) out.push_back(0);
    }
    return true;
}

COBSStreamDecoder::COBSStreamDecoder(size_t max_frame) : max_(max_frame) {}

void COBSStreamDecoder::feed(const uint8_t* data, size_t len) {
    if (!data || len == 0) return;
    for (size_t i = 0; i < len; ++i) {
        uint8_t b = data[i];
        if (b == 0) {
            if (!buf_.empty()) {
                std::vector<uint8_t> decoded;
                if (cobs_decode(buf_, decoded)) {
                    frames_.push_back(std::move(decoded));
                }
            }
            buf_.clear();
        } else {
            if (buf_.size() < max_) buf_.push_back(b); else buf_.clear();
        }
    }
}

bool COBSStreamDecoder::pop_frame(DecodedFrame& outf) {
    if (frames_.empty()) return false;
    std::vector<uint8_t> f = std::move(frames_.front());
    frames_.erase(frames_.begin());
    return parse_frame(f, outf);
}

bool parse_frame(const std::vector<uint8_t>& decoded, DecodedFrame& out) {
    if (decoded.size() < HEADER_SIZE + 2) return false;
    const size_t total = decoded.size();
    // CRC check
    uint16_t rx_crc = static_cast<uint16_t>(decoded[total-2]) |
                      (static_cast<uint16_t>(decoded[total-1]) << 8);
    uint16_t calc = crc16_ccitt(decoded.data(), total - 2);
    if (rx_crc != calc) return false;
    uint8_t msg_id = decoded[0];
    uint8_t len = decoded[1];
    uint8_t seq = decoded[2];
    uint32_t ts = static_cast<uint32_t>(decoded[3]) |
                  (static_cast<uint32_t>(decoded[4]) << 8) |
                  (static_cast<uint32_t>(decoded[5]) << 16) |
                  (static_cast<uint32_t>(decoded[6]) << 24);
    if (HEADER_SIZE + len != total - 2) return false;
    out.msg_id = msg_id;
    out.seq = seq;
    out.ts_ms = ts;
    out.payload.assign(decoded.begin() + HEADER_SIZE, decoded.begin() + HEADER_SIZE + len);
    return true;
}

} // namespace proto

